pkg/clusteragent/admission/patch/patcher.go

```diff
@@ -58,99 +63,152 @@ func (p *patcher) start(stopCh <-chan struct{}) {
 	}
 }
 
-// patchDeployment applies a patch request to a k8s target deployment
-func (p *patcher) patchDeployment(req Request) error {
+func (p *patcher) patchNamespaces(req Request) error {
 	if !p.isLeader() {
 		log.Debug("Not leader, skipping")
 		return nil
 	}
-	deploy, err := p.k8sClient.AppsV1().Deployments(req.K8sTarget.Namespace).Get(context.TODO(), req.K8sTarget.Name, metav1.GetOptions{})
-	if err != nil {
-		return err
-	}
-	oldObj, err := json.Marshal(deploy)
-	if err != nil {
-		return fmt.Errorf("failed to encode object: %v", err)
-	}
-	revision := fmt.Sprint(req.Revision)
-	if deploy.Annotations == nil {
-		deploy.Annotations = make(map[string]string)
-	}
-	if deploy.Annotations[k8sutil.RcIDAnnotKey] == req.ID && deploy.Annotations[k8sutil.RcRevisionAnnotKey] == revision {
-		log.Infof("Remote Config ID %q with revision %q has already been applied to object %s, skipping", req.ID, revision, req.K8sTarget)
-		return nil
+
+	var namespaces []string
+	if req.Action == DeleteConfig {
+		// If the action is to hard delete existing RC configuration, get the list of namespaces that were instrumented by the RC configuration
+		namespaces = *p.configIDToNamespaces[req.ID]
+		defer delete(p.configIDToNamespaces, req.ID)
+	} else {
+		// If the action is to enable/disable RC configuration, get the list of namespaces from K8sTarget
+		namespaces = p.getNamespacesToInstrument(req.K8sTarget.ClusterTargets)
+		p.configIDToNamespaces[req.ID] = &namespaces
 	}
-	log.Infof("Applying Remote Config ID %q with revision %q and action %q to object %s", req.ID, revision, req.Action, req.K8sTarget)
-	switch req.Action {
-	case StageConfig:
-		// Consume the config without triggering a rolling update.
-		log.Debugf("Remote Config ID %q with revision %q has a \"stage\" action. The pod template won't be patched, only the deployment annotations", req.ID, revision)
-	case EnableConfig:
-		if err := enableConfig(deploy, req); err != nil {
+
+	for _, ns := range namespaces {
+		namespace, err := p.k8sClient.CoreV1().Namespaces().Get(context.TODO(), ns, metav1.GetOptions{})
```

Comment:
This is interesting, besides saving the one namespace `get` call, what would we be saving here? I guess if we're doing all of the namespaces this just gets more and more complex-- but then we'd have to maintain a mapping of namespace name to workload id ourselves. 

This also shouldn't happen too often IIRC since this is only for when we get new RC for enabled/disabled namespaces for this specific cluster, right?